components/concept/storage/src/main/java/mozilla/components/concept/storage/BookmarksStorage.kt

```diff
@@ -0,0 +1,82 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.storage
+
+import mozilla.appservices.places.BookmarkFolder
+import mozilla.appservices.places.BookmarkItem
+import mozilla.appservices.places.BookmarkRoot
+import mozilla.appservices.places.BookmarkTreeNode
+import mozilla.appservices.places.BookmarkUpdateInfo
+
+interface BookmarksStorage {
+
+    /**
+     * Produces a populated bookmarks tree for the given root.
+     *
+     * @param root The bookmark root to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root.
+     */
+    suspend fun getTree(root: BookmarkRoot, recursive: Boolean = false): BookmarkTreeNode?
+
+    /**
+     * Produces a populated bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    suspend fun getTree(guid: String, recursive: Boolean = false): BookmarkTreeNode?
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching BookmarkItems.
+     */
+    suspend fun searchBookmarks(query: String, limit: Int = DEFAULT_BOOKMARK_SEARCH_LIMIT): List<BookmarkItem>
+
+    /**
+     * Adds a new bookmark tree node to a given node.
+     *
+     * @param newNode The node to add.
+     * @param position The optional position to add it in.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addNode(target: BookmarkFolder, newNode: BookmarkTreeNode, position: Int? = null): String
+
+    /**
+     * Moves a bookmark tree node to a given node.
+     *
+     * @param newFolder The node to move it to.
+     * @param position The optional position to move it to.
+     * @return The guid of the moved node.
+     */
+    suspend fun moveNode(nodeToMove: BookmarkTreeNode, newFolder: BookmarkFolder, position: Int? = null): String
+
+    /**
+     * Edits an existing bookmark item.
+     *
+     * @param info The info to change in the bookmark.
+     */
+    suspend fun updateNode(node: BookmarkTreeNode, info: BookmarkUpdateInfo)
+
+    /**
+     * Deletes a bookmark node and all of its children, if any.
+     *
+     * @return Whether the bookmark existed or not.
+     */
+    suspend fun deleteNode(node: BookmarkTreeNode): Boolean
+
+    /**
+     * Cleanup all resources used.
+     */
+    fun cleanup()
+
+    companion object {
+        const val DEFAULT_BOOKMARK_SEARCH_LIMIT = 10
```

Comment:
A concrete implementation of this class may be a better place for constants like these.

Reply:
It's a default argument. An overriding function is not allowed to specify default values. Either I must disallow any default value in concrete implementations, I can specify a val which must be overridden, or I specify a default here. Each has upsides and downsides.

Reply:
I decided to supply an empty val that must be overridden. That way the concrete implementation always supplies a default.